tests/test_IDSM.py

```diff
@@ -21,13 +21,15 @@ def test_service_available():
 def test_format():
     query = f"""
     SELECT DISTINCT ?value ?type
+    FROM pubchem:compound FROM pubchem:inchikey FROM descriptor:compound
```

Comment:
Does this mean we are only searching pubchem?

Reply:
Yes, it does. However, the original version of the SPARQL queries also searched only in data originating from PubChem. For searches also in data coming from other sources, the SPARQL queries would have to look different. This is because IDSM integrates data from different sources, but does not unify their structure.

Reply:
The query will be a bit more complicated, because there are several differences between the datasets.

For example, PubChem uses attribute types sio:CHEMINF_000396 (InChI calculated by library version 1.0.4), sio:CHEMINF_000399 (InChIKey generated by software version 1.0.4), sio:CHEMINF_000335 (molecular formula calculated by the pubchem software library), and sio:CHEMINF_000376 (canonical smiles generated by OEChem), but other datasets use attribute types sio:CHEMINF_000113 (InChI descriptor), sio:CHEMINF_000059 (InChIKey), sio:CHEMINF_000042 (molecular formula), and sio:CHEMINF_000018 (SMILES descriptor), respectively.

Another problem is that PubChem stores InChI identifiers as literals with language tags, but other datasets store InChI identifiers as string literals.

For example, a general query retrieving data based on InChI could look like this:

```
SELECT DISTINCT ?value ?type
FROM pubchem:compound FROM pubchem:inchikey FROM descriptor:compound
FROM mona: FROM isdb: FROM ebi:chembl
WHERE
{
  VALUES ?type {
    sio:CHEMINF_000396 sio:CHEMINF_000382
    sio:CHEMINF_000399 sio:CHEMINF_000335
    sio:CHEMINF_000376 sio:CHEMINF_000379
    sio:CHEMINF_000113 sio:CHEMINF_000059
    sio:CHEMINF_000042 sio:CHEMINF_000018 }
  ?compound sio:SIO_000008 [
    rdf:type ?type;
    sio:SIO_000300 ?value ].
  { ?compound sio:SIO_000008 [ rdf:type sio:CHEMINF_000113; sio:SIO_000300 'inchi identifier' ] }
  UNION
  { ?compound sio:SIO_000008 [ rdf:type sio:CHEMINF_000396; sio:SIO_000300 'inchi identifier'@en ] }
}
```

The FROM clauses are there for optimization only. This query includes data from PubChem, ChEMBL, MoNA and ISDB. Nevertheless, it does not include data from ChEBI because ChEBI does not use the SIO ontology at all.

The following query contains data from ChEBI in addition to the previous datasets:

```
SELECT DISTINCT ?value ?type
FROM pubchem:compound FROM pubchem:inchikey FROM descriptor:compound
FROM mona: FROM isdb: FROM ebi:chembl FROM ebi:chebi
WHERE
{
  {
    ?compound sio:SIO_000008 [
      rdf:type sio:CHEMINF_000396;
      sio:SIO_000300 ?value ].
    BIND(sio:CHEMINF_000396 as ?type)
  } UNION {
    ?compound sio:SIO_000008 [
      rdf:type sio:CHEMINF_000382;
      sio:SIO_000300 ?value ].
    BIND(sio:CHEMINF_000382 as ?type)
  } UNION {
    ?compound sio:SIO_000008 [
      rdf:type sio:CHEMINF_000399;
      sio:SIO_000300 ?value ].
    BIND(sio:CHEMINF_000399 as ?type)
  } UNION {
    ?compound sio:SIO_000008 [
      rdf:type sio:CHEMINF_000335;
      sio:SIO_000300 ?value ].
    BIND(sio:CHEMINF_000335 as ?type)
  } UNION {
    ?compound sio:SIO_000008 [
      rdf:type sio:CHEMINF_000376;
      sio:SIO_000300 ?value ].
    BIND(sio:CHEMINF_000376 as ?type)
  } UNION {
    ?compound sio:SIO_000008 [
      rdf:type sio:CHEMINF_000379;
      sio:SIO_000300 ?value ].
    BIND(sio:CHEMINF_000379 as ?type)
  } UNION {
    ?compound sio:SIO_000008 [
      rdf:type sio:CHEMINF_000113;
      sio:SIO_000300 ?value ].
    BIND(sio:CHEMINF_000113 as ?type)
  } UNION {
    ?compound sio:SIO_000008 [
      rdf:type sio:CHEMINF_000059;
      sio:SIO_000300 ?value ].
    BIND(sio:CHEMINF_000059 as ?type)
  } UNION {
    ?compound sio:SIO_000008 [
      rdf:type sio:CHEMINF_000042;
      sio:SIO_000300 ?value ].
    BIND(sio:CHEMINF_000042 as ?type)
  } UNION {
    ?compound sio:SIO_000008 [
      rdf:type sio:CHEMINF_000018 ;
      sio:SIO_000300 ?value ].
    BIND(sio:CHEMINF_000018  as ?type)
  } UNION {
    ?compound chebi:inchikey ?value.
    BIND(sio:CHEMINF_000059 as ?type)
  } UNION {
    ?compound chebi:smiles ?value.
    BIND(sio:CHEMINF_000018 as ?type)
  } UNION {
    ?compound chebi:inchi ?value.
    BIND(sio:CHEMINF_000113 as ?type)
  } UNION {
    ?compound chebi:formula ?value.
    BIND(sio:CHEMINF_000042 as ?type)
  }

  { ?compound sio:SIO_000008 [ rdf:type sio:CHEMINF_000396; sio:SIO_000300 'inchi identifier'@en ] }
  UNION
  { ?compound sio:SIO_000008 [ rdf:type sio:CHEMINF_000113; sio:SIO_000300 'inchi identifier' ] }
  UNION
  { ?compound chebi:inchi 'inchi identifier' }
}
```

The performance of this query is even better than the previous one, but it is not very elegant code.

Reply:
Thanks for this excellent example and explanation!